File: src/Solutions/Euler083.hs
module Solutions.Euler083 where
import Utils.Input(getInput)
import Solutions.Euler081(Matrix, parseMatrix)
import Control.Monad
import Control.Monad.ST
import Data.Array
import Data.Array.ST
import Data.STRef

inf = 10^12
findMinPath :: Matrix -> Int
findMinPath matrix = let n = snd $ snd $ bounds matrix in runST $ do
    dist <- newArray ((0, 0), (n, n)) inf :: ST s (STArray s (Int, Int) Int)
    update <- newArray ((0, 0), (n, n)) False :: ST s (STArray s (Int, Int) Bool)
    writeArray dist (0, 0) $ matrix ! (0, 0)
    writeArray update (0, 0) True
    let selectIndex = do
            selected <- newSTRef ((-1, -1), inf)
            forM_ [0..n] $ \x -> forM_ [0..n] $ \y -> do
                wasUpdated <- readArray update (x, y)
                when wasUpdated $ do
                    (_, oldVal) <- readSTRef selected
                    newVal <- readArray dist (x, y)
                    when (newVal < oldVal) $ writeSTRef selected ((x, y), newVal)
            readSTRef selected
    let loop = do
            ((x, y), d) <- selectIndex
            when ((x, y) /= (-1, -1)) $ do
                writeArray update (x, y) False
                when (x > 0) $ do
                    let left = d + matrix ! (x - 1, y)
                    distLeft <- readArray dist (x - 1, y)
                    when (left < distLeft) $ do
                        writeArray dist (x - 1, y) left
                        writeArray update (x - 1, y) True
                when (x < n) $ do
                    let right = d + matrix ! (x + 1, y)
                    distRight <- readArray dist (x + 1, y)
                    when (right < distRight) $ do
                        writeArray dist (x + 1, y) right
                        writeArray update (x + 1, y) True
                when (y < n) $ do
                    let down = d + matrix ! (x, y + 1)
                    distDown <- readArray dist (x, y + 1)
                    when (down < distDown) $ do
                        writeArray dist (x, y + 1) down
                        writeArray update (x, y + 1) True
                when (y > 0) $ do
                    let up = d + matrix ! (x, y - 1)
                    distUp <- readArray dist (x, y - 1)
                    when (up < distUp) $ do
                        writeArray dist (x, y - 1) up
                        writeArray update (x, y - 1) True
                loop
    loop
    readArray dist (n, n)

euler083 :: IO String
euler083 = show . findMinPath . parseMatrix <$> getInput 83

File: rust/src/solutions/euler083.rs
use crate::solutions::euler081::parse_matrix;
use crate::utils::input::get_input;
use itertools::Itertools;

pub fn euler083() -> String {
    let input = parse_matrix(&get_input(83));
    let l = input.len();
    let mut dist = vec![vec![1000000000usize; l]; l];
    let mut update = vec![vec![false; l]; l];
    dist[0][0] = input[0][0];
    update[0][0] = true;
    loop {
        match (0..l)
            .cartesian_product(0..l)
            .filter(|&(i, j)| update[i][j])
            .min_by_key(|&(i, j)| dist[i][j])
        {
            None => break dist[l - 1][l - 1].to_string(),
            Some((i, j)) => {
                update[i][j] = false;
                if j < l - 1 {
                    let right = dist[i][j] + input[i][j + 1];
                    if right < dist[i][j + 1] {
                        dist[i][j + 1] = right;
                        update[i][j + 1] = true;
                    }
                }
                if j > 0 {
                    let left = dist[i][j] + input[i][j - 1];
                    if left < dist[i][j - 1] {
                        dist[i][j - 1] = left;
                        update[i][j - 1] = true;
                    }
                }
                if i < l - 1 {
                    let down = dist[i][j] + input[i + 1][j];
                    if down < dist[i + 1][j] {
                        dist[i + 1][j] = down;
                        update[i + 1][j] = true;
                    }
                }
                if i > 0 {
                    let up = dist[i][j] + input[i - 1][j];
                    if up < dist[i - 1][j] {
                        dist[i - 1][j] = up;
                        update[i - 1][j] = true;
                    }
                }
            }
        }
    }
}
